torchao/quantization/quant_api.py

```diff
@@ -199,3 +201,382 @@ def replace_conv2d_1x1(conv):
+            qparams: the output from get_qparams_func
+        Returns:
+            weight: A 2d weight tensor with non-integer dtype.
+
+    combine_qparams_list_func:
+        A function that combines several qparams into one qparam.
+        Args:
+            qparams_list: a list of qparams objects, each obtained by calling get_qparams_func
+            on a single group from a weight tensor
+        Returns:
+            qparams: an object of the same format as the qparams above.
+
+    skip_layer_func:
+        A function that determines which linear layers should be skipped during GPTQ
+        Args:
+            weight: A 2d weight tensor with non-integer dtype.
+        Returns:
+            skip: boolean indicating whether layer should be skipped
+
+    make_names_and_values_dict_func:
+        A function that prepares the qparams and quantized_weight and creates a dictionary indicating how they
+        should be inserted into the state_dict. Generally any packing of the weight and qparams should be done here.
+        Args:
+            quantized_weight: A 2d quantized weight tensor (generally with an integer dtype)
+            qparams: the output from get_qparams_func
+        Returns:
+            names_and_values_dict: a dictionary mapping the name of the parameters of the quantized module to the
+            corresponding quantized weights and qparams.
+    """
+
+    def __init__(self):
+        assert self.get_qparams_func is not None
+        assert self.quantize_func is not None
+        assert self.dequantize_func is not None
+        assert self.combine_qparams_list_func is not None
+        assert self.make_names_and_values_dict_func is not None
+
+    @staticmethod
+    def get_inputs(
+        model,
+        tokenizer,
+        calibration_tasks,
+        calibration_limit,
+        calibration_seq_length,
+        pad_calibration_inputs,
+    ) -> "MultiInput":  # pyre-ignore[11]
+        input_recorder = InputRecorder(
+            model,
+            tokenizer,
+            calibration_seq_length,
+            pad_calibration_inputs,
+        )
+
+        try:
+            lm_eval.tasks.initialize_tasks()
+        except:
+            pass
+        task_dict = get_task_dict(calibration_tasks)
+        print("Obtaining GPTQ calibration inputs on: ", calibration_tasks)
+
+        evaluate(
+            input_recorder,
+            task_dict,
+            limit=calibration_limit,
+        )
+        inputs = input_recorder.get_recorded_inputs()
+        assert inputs is not None, (
+            f"No inputs were collected, use a task other than {calibration_tasks}, "
+            + "use option pad_calibration_inputs, or decrease calibration_sequence_length (currently "
+            + f"{calibration_seq_length})"
+        )
+        print(f"Obtained {len(inputs[0].values)} calibration samples")
+        return inputs
+
+    @torch.no_grad()
+    def _create_quantized_state_dict(
```

Comment:
Why does this need this private method and then also a public `quantize`?

Reply:
Current quantization, save/load flow is a bit unintuitive: https://github.com/pytorch/executorch/blob/main/examples/models/llama2/export_llama_lib.py#L226-L237
since people need to call these special functions.

I want `quantize` to return a real quantized model with modified weights so that people can save the model the normal way (torch.save(quantized_model.state_dict())

see the one step quant flow:
```
# One Step Quantizer Flow
# 1. initialize quantizer
quantizer = MyQuantizer(...configs)
# 2. quantize the model and also convert the state_dict
model = quantizer.quantize(model)

# 3. save/load the prepared model state (with state_dict())
torch.save(model.state_dict(), “file_name.pt”)

# for load we’ll need to prepare the model again
quantizer = MyQuantizer(...configs)
model = quantizer.quantize(model)
model.load_state_dict(torch.load(“file_name.pt”))

# 4. lower to model to some backend, e.g. torch.compile, executorch
model = lower_to_backendA(model)
```